src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,155 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode;
+        if (first == null) {
+            newNode = new Node<>(null, value, null);
+            first = newNode;
+        } else {
+            newNode = new Node<>(last, value, null);
+            last.next = newNode;
+        }
+        last = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkElementIndex(index, size + 1);
+        if (index == size) {
+            add(value);
+        } else {
+            addBefore(value, getNodeByIndex(index));
+        }
+    }
+
+    private void addBefore(T value, Node<T> nodeByIndex) {
+        Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+        if (nodeByIndex.prev != null) {
+            nodeByIndex.prev.next = newNode;
+        } else {
+            first = newNode;
+        }
+        nodeByIndex.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index, size);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index, size);
+        Node<T> current = getNodeByIndex(index);
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index, size);
+        return unlink(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> node = first; node != null; node = node.next) {
+                if (node.value == null) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> node = first; node != null; node = node.next) {
+                if (object == node.value || object.equals(node.value)) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
```

Comment:
won't be visible outside of this class anyway
```suggestion
        Node(Node<T> prev, T value, Node<T> next) {
```